bot/exts/moderation/metabase.py

```diff
@@ -0,0 +1,173 @@
+
+        # Cache the session info, since login in heavily ratelimitted
+        await self.session_info.set("session_token", self.session_token)
+        await self.session_info.set("session_expiry", refresh_time.timestamp())
+
+        self._session_scheduler.schedule_at(refresh_time, 0, self.refresh_session())
+
+    @group(name="metabase", invoke_without_command=True)
+    async def metabase_group(self, ctx: Context) -> None:
+        """A group of commands for interacting with metabase."""
+        await ctx.send_help(ctx.command)
+
+    @metabase_group.command(name="extract")
+    async def metabase_extract(self, ctx: Context, question_id: int, extension: Optional[str] = "csv") -> None:
+        """
+        Extract data from a metabase question.
+
+        You can find the question_id at the end of the url on metabase.
+        I.E. /question/{question_id}
+
+        If, instead of an id, there is a long URL, make sure to save the question first.
+
+        If you want to extract data from a question within a dashboard, click the
+        question title at the top left of the chart to go directly to that page.
+
+        Valid extensions are: csv and json.
+        """
+        async with ctx.typing():
+
+            # Make sure we have a session token before running anything
+            await self.init_task
+
+            if extension not in ("csv", "json"):
+                # "api" and "xlsx" are supported by metabase's api, but wouldn't work for exporting to pastebin
+                raise BadArgument(f"{extension} is not a valid extension!")
+
+            url = f"{MetabaseConfig.url}/card/{question_id}/query/{extension}"
+            async with self.bot.http_session.post(url, headers=self.headers) as resp:
+                try:
+                    resp.raise_for_status()
+                except ClientResponseError as e:
+                    if e.status == 403:
+                        # User doesn't have access to the given question
+                        log.warn(f"Failed to auth with Metabase for question {question_id}.")
+                        await ctx.send(f":x: {ctx.author.mention} Failed to auth with Metabase for that question.")
+                    else:
+                        # User credentials are invalid, or the refresh failed.
+                        # Delete the expiry time, to force a refresh on next startup.
+                        await self.session_info.delete("session_expiry")
+                        log.exception("Session token is invalid or refresh failed.")
+                        await ctx.send(f":x: {ctx.author.mention} Session token is invalid or refresh failed.")
+                    return
+
+                if extension == "csv":
+                    out = await resp.text()
+                    # Save the output for use with int e
+                    with StringIO(out) as f:
+                        self.exports[question_id] = list(csv.DictReader(f))
```

Comment:
```suggestion
                    self.exports[question_id] = list(csv.DictReader(StringIO(out)))
```
Don't really need a context manager for `StringIO` as it has no underlying connections